File: ScaleformUI_Csharp/Items/UIMenuDynamicListItem.cs
using CitizenFX.Core.UI;
using System;
using System.Drawing;
using System.Threading.Tasks;
using Font = CitizenFX.Core.UI.Font;

namespace ScaleformUI
{
    public class UIMenuDynamicListItem : UIMenuItem, IListItem
    {
        public enum ChangeDirection
        {
            Left,
            Right
        }

        public delegate Task<string> DynamicListItemChangeCallback(UIMenuDynamicListItem sender, ChangeDirection direction);
        private string currentListItem;

        public string CurrentListItem
        {
            get => currentListItem; internal set
            {
                currentListItem = value;
                if(Parent is not null && Parent.Visible)
                {
                    ScaleformUI._ui.CallFunction("UPDATE_LISTITEM_LIST", Parent.MenuItems.IndexOf(this), currentListItem, 0);
                }
            }
        }
        public DynamicListItemChangeCallback Callback { get; set; }

        /// <summary>
        /// List item with items generated at runtime
        /// </summary>
        /// <param name="text">Label text</param>
        public UIMenuDynamicListItem(string text, string startingItem, DynamicListItemChangeCallback changeCallback) : this(text, null, startingItem, changeCallback)
        {
        }

        /// <summary>
        /// List item with items generated at runtime
        /// </summary>
        /// <param name="text">Label text</param>
        /// <param name="description">Item description</param>
        public UIMenuDynamicListItem(string text, string description, string startingItem, DynamicListItemChangeCallback changeCallback) : base(text, description)
        {
            _itemId = 1;
            currentListItem = startingItem;
            Callback = changeCallback;
        }

        public override void SetRightBadge(BadgeIcon badge)
        {
            throw new Exception("UIMenuListItem cannot have a right badge.");
        }

        public override void SetRightLabel(string text)
        {
            throw new Exception("UIMenuListItem cannot have a right label.");
        }

        public string CurrentItem()
        {
            return CurrentListItem;
        }
    }
}

File: ScaleformUI_V8/src/menus/UIMenu/items/uimenudynamiclistitem.ts
import { UIMenuItem } from "./uimenuitem";
import { UIMenuCheckboxItemChangeCallback, UIMenuDynamicListItemChangeCallbackBuilder } from "../emitters/emitters";
import { BadgeStyle } from "elements/badge";

export enum ChangeDirection {
    Left = 0,
    Right,
}

export class UIMenuDynamicListItem extends UIMenuItem {
    callback = new UIMenuDynamicListItemChangeCallbackBuilder();
    private currentListItem: string = "";

    public set CurrentListItem(value: string) {
        this.currentListItem = value;
        if (this.Parent !== null && this.Parent.Visible && this.Parent.Pagination.IsItemVisible(this.Parent.Items.indexOf(this))) {
            ScaleformUI.Scaleforms._ui.callFunction("UPDATE_LISTITEM_LIST", this.Parent.Pagination.GetScaleformIndex(this.Parent.Items.indexOf(this)),
                this.currentListItem, 0);
        }
    }
    public get CurrentListItem(): string {
        return this.currentListItem;
    }

    constructor(text: string, description: string, startingItem: string, changeCallback: UIMenuCheckboxItemChangeCallback) {
        super(text, description);
        this._itemId = 1;
        this.currentListItem = startingItem;
        this.callback.add(changeCallback);
    }

    public SetRightBadge(badge: BadgeStyle) {
        throw new Error("UIMenuDynamicListItem cannot have a right badge.");
    }

    public SetRightLabel(text: string) {
        throw new Error("UIMenuDynamicListItem cannot have a right label.");
    }
}